build-tools/xaprepare/xaprepare/Steps/Step_DetermineAzurePipelinesTestJobs.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Xamarin.Android.Prepare
+{
+	class Step_DetermineAzurePipelinesTestJobs : Step
+	{
+		public Step_DetermineAzurePipelinesTestJobs ()
+			: base ("Determine Azure Pipelines test jobs to run from a merge commit")
+		{}
+
+		protected override async Task<bool> Execute (Context context)
+		{
+			Log.StatusLine ("Determining test jobs to run...");
+
+			string commitRevision = Environment.GetEnvironmentVariable ("BUILD_SOURCEVERSION");
+			string commitMessage = Environment.GetEnvironmentVariable ("BUILD_SOURCEVERSIONMESSAGE");
+			if (string.IsNullOrEmpty (commitRevision) || string.IsNullOrEmpty (commitMessage)) {
+				Log.ErrorLine ("One or more source version variable values were empty:");
+				Log.ErrorLine ($"BUILD_SOURCEVERSION='{commitRevision}' BUILD_SOURCEVERSIONMESSAGE='{commitMessage}'.");
+				return false;
+			}
+
+			// Assume we're building a merge commit as part of an Azure Pipelines PR build. Otherwise, this step will fail.
+			//  Example: Merge 0b66502c8b9f33cbb8d21b2dab7c100629aec081 into 0bef8aa5cd74d83d77c4e2b3f63975a0deb804b3
+			var commitMessagePieces = commitMessage.Split (new string [] { " " }, StringSplitOptions.RemoveEmptyEntries);
+			if (string.IsNullOrEmpty (commitMessagePieces [3])) {
+				Log.ErrorLine ($"Unable to parse merge commit message from: '{commitMessage}'.");
+				return false;
+			}
+
+			var git = new GitRunner (context);
+			var filesChanged = await git.RunCommandForOutputAsync (BuildPaths.XamarinAndroidSourceRoot, "diff", "--name-only", commitRevision, commitMessagePieces [3]);
+			if (filesChanged == null || filesChanged.Count < 1) {
+				Log.ErrorLine ($"Unable to determine if any files were changed in this PR.");
+				return false;
+			}
+
+			var testAreas = new List<string> ();
+			foreach (string file in filesChanged) {
+				// Compare files changed against common areas requiring additional test scope.
+				// MSBuild: Runs all legacy and One .NET Xamarin.Android.Build.Task tests, as well as designer tests.
+				// Mono: Runs all BCL and timezone unit tests, as well as designer tests.
+				if (file == ".external") {
+					testAreas.Add ("Mono");
+					testAreas.Add ("MSBuild");
+				} else if (file.Contains ("external/Java.Interop")) {
+					testAreas.Add ("MSBuild");
+				} else if (file.Contains ("Xamarin.Android.Build.Tasks")) {
+					testAreas.Add ("MSBuild");
+				}
```

Comment:
One missing here is if you modified `src/monodroid`: I think @grendello would want the designer tests to run. But maybe doesn't need the BCL tests in that case?

Reply:
Yep, BCL tests wouldn't be necessary. Designer tests, however, need to be ran for sure (and possibly the XF integration test)

Reply:
It would be a good idea to check for the test paths themselves - if a test is modified by the commit, it should be ran.

Reply:
Then, `Mono.Android` changes should trigger `Mono.Android/Test` + its 2 derivatives (MultiDex, AppBundle)

Reply:
Then any Android NDK or SDK updates should trigger all the tests, IMO

Reply:
All of the APK tests (including XF integration) will continue to run against every PR. That job executes in under one hour and I felt that we didn't need to cut scope there at this time.

I don't want to overcomplicate this by introducing source checks and job conditions for every test suite that we have. If an author is modifying a test suite that didn't run as part of the PR build it should hopefully be clear to the author that the relevant stage needs to be manually queued. I think the same applies for NDK/SDK updates.

I'd like this PR to cover the most common source areas that authors may be uncertain about relevant test coverage for, not every source area.

Reply:
I reworked this to use a single variable that will contain multiple test area values. Valid values are associated with test job types instead of test suites.